pkg/app/api/grpcapi/piped_api.go

```diff
@@ -957,8 +957,75 @@ func (a *PipedAPI) GetDesiredVersion(ctx context.Context, _ *pipedservice.GetDes
 }
 
 func (a *PipedAPI) UpdateApplicationConfigurations(ctx context.Context, req *pipedservice.UpdateApplicationConfigurationsRequest) (*pipedservice.UpdateApplicationConfigurationsResponse, error) {
-	// TODO: Update the given application configurations
-	return nil, status.Errorf(codes.Unimplemented, "UpdateApplicationConfigurations is not implemented yet")
+	projectID, pipedID, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Consider bulk-updating multiple apps
+	for _, appInfo := range req.Applications {
+		opts := datastore.ListOptions{
+			// NOTE: Assume that no more than one application is referring to a single configuration file.
+			Limit: 1,
+			Filters: []datastore.ListFilter{
+				{
+					Field:    "ProjectId",
+					Operator: datastore.OperatorEqual,
+					Value:    projectID,
+				},
+				{
+					Field:    "PipedId",
+					Operator: datastore.OperatorEqual,
+					Value:    pipedID,
+				},
+				{
+					Field:    "GitPath.Repo.Id",
+					Operator: datastore.OperatorEqual,
+					Value:    appInfo.RepoId,
+				},
+				{
+					Field:    "GitPath.Path",
+					Operator: datastore.OperatorEqual,
+					Value:    appInfo.Path,
+				},
+				{
+					Field:    "GitPath.ConfigFilename",
+					Operator: datastore.OperatorEqual,
+					Value:    appInfo.ConfigFilename,
+				},
+				{
+					Field:    "Disabled",
+					Operator: datastore.OperatorEqual,
+					Value:    false,
+				},
+			},
+		}
```

Comment:
Application ID can be included in the request by Piped to avoid this listing.

Reply:
Do you mean to uniquely identify the application based on the RepoID, Path, and ConfigFilename, get the id, and cache it on the Piped side?

Reply:
No. I think since this is for updating the **existing applications** then Piped already knew their ID before sending this `UpdateApplicationConfigurations` request.

https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/appconfigreporter/appconfigreporter.go#L146
https://github.com/pipe-cd/pipe/blob/master/pkg/app/piped/appconfigreporter/appconfigreporter.go#L197

So I meant that we can add an `application_id` field to this https://github.com/pipe-cd/pipe/blob/master/pkg/model/common.proto#L59-L65 model or the RPC request to avoid re-finding each one on the server to update.

Reply:
More specifically, we can update this map to be a map from GitPathKey to AppID (`make(map[string]string, len(apps))`)

https://github.com/pipe-cd/pipe/blob/ce3e2b322cb5caef628d7ae7188396eb961beb7a/pkg/app/piped/appconfigreporter/appconfigreporter.go#L147

Reply:
Sounds nice. I originally applied in this way this because I was trying to allow multiple applications to refer to a single config file, but now that the premise has changed, it seems better to identify them by ID.